<h1>Q-Learning</h1>

<p>Στην συγκεκριμένη άσκηση θα μελετήσετε το στοχαστικό αλγόριθμο Q-Learning, χρησιμοποιώντας το έτοιμο πρόγραμμα που σας δίνεται.</p> <p> Στα πλαίσια του παραδείίγματος θα εξετάσετε μίία υλοποίηση με q-learning σχετικά με το σύστημα αυτόματης οδήγησης ενός ταξί. Στα πλαίσια του προβλήματος αυτού θα πρέπει να ικανοποιούνται τα εξής:</p>
<ul>
<li>Το ταξί θα πρέπει να αφήνει τον πελάτη στη σωστή θέση</li>
<li>Το ταξί να ακολουθεί τη συντομότερη δυνατή διαδρομή</li>
<li>Να τηρούνται οι κανόόνες κυκλοφορίας και ασφάλειας των επιβατών</li>
</ul>

<p>Στα πλαίσια του προβλήματος έχουμε τα εξής χαρακτηριστικά για την επιβράβευση, τις καταστάσεις και τις ενέργειες.</p> 

<h4>Επιβράβευση</h4>
<ul>
<li>Θα έχουμε τη μέγιστη επιβράβευση όταν το ταξί αφήνει έναν πελάτη στην επιθυμητή θέση</li>
<li>Θα υπάρχει ποινή στην περίπτωση όπου το ταξί αφήσει τον πελάτη σε κάποιο λανθασμένο σημείο</li>
<li>Ο agent θα παίρνει μία μικρή σχετικά ποινή στην περίπτωση όπου αργεί να φτάσει στον τελικό προορισμό</li>
</ul>

<p>Γενικά οι παραπάνω αρχές συνοψίζονται στα εξής: "Λαμβάνουμε +20 πόντους για μια επιτυχημένη πτώση και χάνουμε 1 πόντο για κάθε χρονικό βήμα που παίρνει. Υπάρχει επίσης ποινή 10 πόντων για παράνομες ενέργειες παραλαβής και αποχώρησης."</p>

<h4>Πλήθος Καταστάσεων</h4>
<img src="https://storage.googleapis.com/lds-media/images/Reinforcement_Learning_Taxi_Env.width-1200.png">
<p>Στο παράδειγμά μας έχουμε ένα μικρό χώρο 5*5. Από εκεί και πέρα έχουμε 4 προορισμούς και 5 πιθανές θέσεις του πελάτη (παίρνουμε και την περίπτωση ο πελάτης να είναι ήδη μέσα στο τάξι).</p>
<p>Με βάση τα παραπάνω έχουμε 5*5*5*4=500 πιθανές καταστάσεις.</p>

<h4>Πλήθος Ενεργειών</h4>
<p>Έχουμε έξι ενεργειες για το ταξί, οι οποίες είναι οι εξής:</p>
<ul>
<li>0=Νότια</li>
<li>1=Βόρεια</li>
<li>2=Ανατολικά</li>
<li>3=Δυτικά</li>
<li>4=Επιβίβαση</li>
<li>5=Αποβίβαση</li>
</ul>





<p>Πριν συνεχίσετε στην άσκηση να απαντήσετε στο εξής ερώτημα:</p>

<b><p>1. Να περιγράψετε σύντομα τον αλγόριθμο Q-Learning. Σε ποια προβλήματα θεωρείτε ότι ταιριάζει ως τρόπος εκμάθησης η Ενισχυτική Μάθηση (Reinforcement Learning); Ποια είναι η βασική διαφορά του αλγορίθμου Q-Learning από τους αλγορίθμους Policy Iteration και Value Iteration;</p></b>

O Αλγόριθμος Q-learning ειναι μια τεχνική της ενισχυστικής μάθησης, που λειτουργεί εκτιμώνατας τη συνάρτηση Q(S,A) για να μάθει μια καλή ή βέλτιστη πολιτική για τον κόσμο του πράκτορα χωρίνα να ξερει το μοντέλο του κόσμου (model-free). Μια βασική προυπόθεση για την εφαρμογή του αλγορίθμου είναι να μπντελοποιείται ο κόσμπς ώς Markov Process (η επόμνη κατάσταση ανεξάρτητη απο τό ιστορικό των κινήσεων και εξαρτάται μονάχα απο την τορινή κτάσταση).Ξεκινώντας απο την τωρινή κατάσταση ο αλγόριθμος βρίσκει την βέλτιστη πολιτική μεγιστοποιώντας την ανταμοιβή σε όλα συνολικά τα βήματα που επιλέγει. 
Προβλήματα στα οποία ζητείται να βρεθεί η βέλτιστη διαδρομή, προβήματα γράφων, ή ακόμα και σε προβλήματα classification είναι αυτά στα οποία ο συγκεκριμένος αλγόριθμος βρίσκει εφαρμογή. 
Η διαφορά με τους αλγορίθμους Policy & Value Iteration είναι οτι εκ των προτερων οι πιθανότητες μετάβασης αλλα και τα οφέλη είναι άγνωστα- δεν υπάρχει συγκεκριμένο μοντέλο που ακολουθείται. Μετά απο την πραγματοποίηση μιας κίνησης του agent γνωστοποιείται η ανταμοιβή για την συγκεκριμένη μετάβαση. Οι μελλοντικές μεταβάσεις γνωστοποιούντια μονάχα όταν ο agend πρέπει να πάρει μια απόφαση απο την συγκεκριμένη θέση. Για στοχαστικές μεταβάσεις οι ο αλγόριθμος εκπαιδεύεται πανω στις πιθανότητες μετάβασης από την επαναληψημότητα μιας κατάστασης. 

<p>Στη συνέχεια θα πρέπει να φορτώσετε τη βιβλιοθήκη gym καθώς και το σχετικό dataset<p>

In [1]:
!pip install cmake 'gym[atari]' scipy

     |████████████████████████████████| 19.4 MB 4.4 MB/s eta 0:00:01
     |████████████████████████████████| 2.8 MB 27.9 MB/s eta 0:00:01


In [2]:
import gym

env = gym.make("Taxi-v3").env

env.render()

+---------+
|R: | : :G|
| : | : : |
| : : : : |
| | : | : |
|Y| : |B: |
+---------+



In [3]:
env.reset() # reset environment to a new, random state
env.render()

print("Action Space {}".format(env.action_space))
print("State Space {}".format(env.observation_space))


+---------+
|R: | : :G|
| : | : : |
| : : : : |
| | : | : |
|Y| : |B: |
+---------+

Action Space Discrete(6)
State Space Discrete(500)


<p>Παρακάτω ορίζουμε τις συνεταγμένες του ταξί, τη θέση του πελάτη και το σημείο προορισμού</p>

In [5]:
state = env.encode(3, 1, 2, 0) # (taxi row, taxi column, passenger index, destination index)
print("State:", state)

env.s = state
env.render()

State: 328
+---------+
|R: | : :G|
| : | : : |
| : : : : |
| | : | : |
|Y| : |B: |
+---------+



<p>Παρακάτω είναι η μήτρα επιβράβευσης για το state που ορίσαμε στο προηγούμενο βήμα</p>

In [6]:
env.P[328]

{0: [(1.0, 428, -1, False)],
 1: [(1.0, 228, -1, False)],
 2: [(1.0, 348, -1, False)],
 3: [(1.0, 328, -1, False)],
 4: [(1.0, 328, -10, False)],
 5: [(1.0, 328, -10, False)]}

<p> Τρέχουμε το παράδειγμά μας χωρις τη χρήση Q-Learning.</p>

<b><p>2. Τα αποτελέσματα είναι ικανοποιητικά; Πως θα μας εξυπηρετούσε η χρήση του Q-Learning;</p></b>

Δεν οδηγούμαστε σε ικανοποιητικά αποτελέσματα καθώς η επιλογή κατάστασης δεν βασίζεται σε penalties ή rewards παρα είναι τυχάια.
Αν χρησιμοποιούσαμε Q-Learning ο agent θα μάθαινε μεσω της επαναληψιμότητας τα rewards μα και τα penalties ειδικότερα τα οποία και θα απέφευγε. 

In [7]:
env.s = 328  # set environment to illustration's state

epochs = 0
penalties, reward = 0, 0

frames = [] # for animation

done = False

while not done:
    action = env.action_space.sample()
    state, reward, done, info = env.step(action)

    if reward == -10:
        penalties += 1
    
    # Put each rendered frame into dict for animation
    frames.append({
        'frame': env.render(mode='ansi'),
        'state': state,
        'action': action,
        'reward': reward
        }
    )

    epochs += 1
    
    
print("Timesteps taken: {}".format(epochs))
print("Penalties incurred: {}".format(penalties))

Timesteps taken: 2147
Penalties incurred: 724


<p>Τώρα θα προσπαθήσουμε να λύσουμε το πρόβλημά μας με τη χρήση του Q-Learning.</p>

<b><p>3. Τι γνωρίζετε για τις παραμέτρους α και γ. Τι θα συμβεί αν έχουν τιμές ίσες με 1;</p></b>

**alpha:** αναφέρεται στο ποσοστό εκμάθησης. Για στοχαστικές συναρτήσεις (ανταμοιβής / μετάβασης) η τιμή του θα πρέπει να είναι φθίνουσα με την αυξηση του αριθμου των επαναλήψεων.Το παραπάνω έχει άμεση συσχέτιση με την προσέγγιση του αναμενόμενου αποτελέσματος (Μεταβαση * Αντμοιβή) για τυχαόία συμπεριφορά τους. Η τιμή 1 αντιστοιχεί σε πολύ γρήγορη ανανέωση των q-values ενώ η τιμή 0 se καθόλου ενημέρωση ( αδυναμία μάθησης)

**gamma:** ανταποκρίνεται στην αξια της ανταμοιβής. Μπορεί να είναι στατική ή δυναμική. Για την τιμή 1 έιναι σημαντική η επιβράβευση όχι μόνο για την άμεση κατάσταση αλλα και για μεταγενέστερες καταστάσεις (ελευθερια επιλογήσ της τωρινής κατάστασης). Τιμή 0 η επιβράβευση είναι σημαντική μονάχα για τις πιο αμεσες ανταμοιβές.

In [8]:
import numpy as np
q_table = np.zeros([env.observation_space.n, env.action_space.n])

In [9]:
%%time
"""Training the agent"""

import random
from IPython.display import clear_output

# Hyperparameters
alpha = 0.1
gamma = 0.6
epsilon = 0.1

# For plotting metrics
all_epochs = []
all_penalties = []

for i in range(1, 100001):
    state = env.reset()

    epochs, penalties, reward, = 0, 0, 0
    done = False
    
    while not done:
        if random.uniform(0, 1) < epsilon:
            action = env.action_space.sample() # Explore action space
        else:
            action = np.argmax(q_table[state]) # Exploit learned values

        next_state, reward, done, info = env.step(action) 
        
        old_value = q_table[state, action]
        next_max = np.max(q_table[next_state])
        
        new_value = (1 - alpha) * old_value + alpha * (reward + gamma * next_max)
        q_table[state, action] = new_value

        if reward == -10:
            penalties += 1

        state = next_state
        epochs += 1
        
    if i % 100 == 0:
        clear_output(wait=True)
        print(f"Episode: {i}")

print("Training finished.\n")

Episode: 100000
Training finished.

CPU times: user 1min 14s, sys: 22.2 s, total: 1min 37s
Wall time: 1min 13s


In [10]:
q_table[328]

array([ -2.41329978,  -2.27325184,  -2.41590663,  -2.35906783,
       -10.68215401, -10.09079863])

In [15]:
"""Evaluate agent's performance after Q-learning"""

total_epochs, total_penalties,total_rewards = 0, 0, 0
episodes = 100

for _ in range(episodes):
    state = env.reset()
    epochs, penalties, reward, rewards = 0, 0, 0, 0
    
    done = False
    
    while not done:
        action = np.argmax(q_table[state])
        state, reward, done, info = env.step(action)

        if reward == -10:
            penalties += 1
        else:
            rewards +=1

        epochs += 1

    total_penalties += penalties
    total_epochs += epochs
    total_rewards +=rewards
    
print(f"Results after {episodes} episodes:")
print(f"Average timesteps per episode: {total_epochs / episodes}")
print(f"Average penalties per episode: {total_penalties / episodes}")
print(f"Average rewaards per episode: {total_rewards / episodes}")

Results after 100 episodes:
Average timesteps per episode: 13.34
Average penalties per episode: 0.0
Average rewaards per episode: 13.34


In [16]:
"""Evaluate agent's performance without Q-learning"""
env.reset() # reset environment to a new, random state

env.s = 328  # set environment to illustration's state

episodes = 100

frames = [] # for animation
total_epochs = 0
total_penalties = 0
total_rewards = 0
done = False
for _ in range(episodes):
  state = env.reset()
  epochs, penalties, reward, rewards = 0, 0, 0, 0
  while not done:
      action = env.action_space.sample()
      state, reward, done, info = env.step(action)

      if reward == -10:
          penalties += 1
      else:
          rewards += 1
      
      # Put each rendered frame into dict for animation
      frames.append({
          'frame': env.render(mode='ansi'),
          'state': state,
          'action': action,
          'reward': reward
          }
      )

      epochs += 1
  total_epochs += epochs
  total_penalties += penalties
  total_rewards += rewards
    
print(f"Results after {episodes} episodes:")
print(f"Average timesteps per episode: {total_epochs / episodes}")
print(f"Average penalties per episode: {total_penalties / episodes}")
print(f"Average rewards per episode: {total_rewards / episodes}")

Results after 100 episodes:
Average timesteps per episode: 9.68
Average penalties per episode: 2.9
Average rewards per episode: 6.78


<b><p>4. Συγκρίνετε τους δύο αλγορίθμους με βάση τις παρακάτω μετρικές</p>
<ul>
<li>Μέσος αριθμός παραβάσεων ανά επεισόδιο</li>
<li>Μέσος αριθμός βημάτων ανά διαδρομή</li>
<li>Μέσος αριθμός ανταμοιβών ανά κίνηση</li>
</ul>
<p>Τις παραπάνω συγκρίσεις να τις κάνετε για 100 επεισόδια.</p>
</b>

Με την χρηση Q-Learing καταλήγουμε σε καλύτερα αποτελέσματα όπως μπορούμε να δούμε στην εκτέλεση των παραπάνω.

#### Me Q-Learning
Results after 100 episodes:

- Average timesteps per episode: 13.34

- Average penalties per episode: 0.0

- Average rewaards per episode: 13.34

#### Xωρίς Q-Learnign
Results after 100 episodes:

- Average timesteps per episode: 9.68

- Average penalties per episode: 2.9

- Average rewards per episode: 6.78
